File: src/types/tool.rs
//! Tool-related types

use serde::{Deserialize, Serialize};

/// ツール情報（一覧表示用）
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ToolInfo {
    pub name: String,
    pub display_name: String,
    pub description: String,
    pub version: String,
    pub icon: Option<String>,
    pub category: Option<String>,
    pub tool_dir: String,
}

/// ツール設定（詳細用）
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ToolConfig {
    pub name: String,
    pub display_name: String,
    pub description: String,
    pub version: String,
    pub binary: String,
    pub icon: Option<String>,
    pub category: Option<String>,
    pub options: Vec<ToolOption>,
    pub result_parser: Option<ResultParser>,
}

/// コマンドラインオプションの定義
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ToolOption {
    pub name: String,
    pub flag: String,
    pub short_flag: Option<String>,
    #[serde(rename = "type")]
    pub option_type: OptionType,
    pub description: String,
    #[serde(default)]
    pub required: bool,
    pub default: Option<serde_json::Value>,
    pub placeholder: Option<String>,
    pub options: Option<Vec<String>>,
    /// パスの種類 ("file", "directory", "any") - Pathタイプの場合のみ使用
    pub path_type: Option<String>,
}

/// オプションの型
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum OptionType {
    String,
    Path,
    Boolean,
    Select,
    Number,
}

/// 結果パーサーの設定
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ResultParser {
    #[serde(rename = "type")]
    pub parser_type: ParserType,
    pub output_flag: Option<String>,
    pub schema: Option<ResultSchema>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum ParserType {
    Json,
    Text,
}

/// 結果のスキーマ定義
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResultSchema {
    pub summary: Option<Vec<SummaryItem>>,
    pub details: Option<DetailsConfig>,
}

/// サマリー項目の定義
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SummaryItem {
    pub key: String,
    pub label: String,
    pub path: String,
    pub count_type: Option<String>,
    pub color: Option<String>,
    pub icon: Option<String>,
}

/// 詳細表示の設定
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DetailsConfig {
    pub items: String,
    pub columns: Vec<ColumnConfig>,
}

/// カラム設定
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ColumnConfig {
    pub key: String,
    pub label: String,
    pub width: Option<String>,
    pub flex: Option<i32>,
}

/// ツール実行結果
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ToolResult {
    pub success: bool,
    pub exit_code: Option<i32>,
    pub stdout: String,
    pub stderr: String,
    pub parsed_result: Option<serde_json::Value>,
}

/// リアルタイムログイベント
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LogEvent {
    pub tool_name: String,
    pub line: String,
    pub stream: LogStream,
    pub timestamp: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum LogStream {
    Stdout,
    Stderr,
}

/// ツール実行状態イベント
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ToolStatusEvent {
    pub tool_name: String,
    pub status: ToolStatus,
    pub result: Option<ToolResult>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum ToolStatus {
    Running,
    Completed,
    Failed,
    Cancelled,
}

/// ログエントリ
#[derive(Debug, Clone)]
pub struct LogEntry {
    pub line: String,
    pub stream: LogStream,
    pub timestamp: String,
}

File: src/types/tool.ts
// Tool-related types

/// オプションの型
export type OptionType = 'string' | 'path' | 'boolean' | 'select' | 'number';

/// ツール情報（一覧表示用）
export interface ToolInfo {
  name: string;
  displayName: string;
  description: string;
  version: string;
  icon: string | null;
  category: string | null;
  toolDir: string;
}

/// ツール設定（詳細用）
export interface ToolConfig {
  name: string;
  displayName: string;
  description: string;
  version: string;
  binary: string;
  icon: string | null;
  category: string | null;
  options: ToolOption[];
  resultParser: ResultParser | null;
}

/// コマンドラインオプションの定義
export interface ToolOption {
  name: string;
  flag: string;
  shortFlag: string | null;
  type: OptionType;
  description: string;
  required: boolean;
  default: unknown | null;
  placeholder: string | null;
  options: string[] | null;
  /// パスの種類 ("file", "directory", "any") - Pathタイプの場合のみ使用
  pathType: string | null;
}

/// パーサーの型
export type ParserType = 'json' | 'text';

/// 結果パーサーの設定
export interface ResultParser {
  type: ParserType;
  outputFlag: string | null;
  schema: ResultSchema | null;
}

/// 結果のスキーマ定義
export interface ResultSchema {
  summary: SummaryItem[] | null;
  details: DetailsConfig | null;
}

/// サマリー項目の定義
export interface SummaryItem {
  key: string;
  label: string;
  path: string;
  countType: string | null;
  color: string | null;
  icon: string | null;
}

/// 詳細表示の設定
export interface DetailsConfig {
  items: string;
  columns: ColumnConfig[];
}

/// カラム設定
export interface ColumnConfig {
  key: string;
  label: string;
  width: string | null;
  flex: number | null;
}

/// ツール実行結果
export interface ToolResult {
  success: boolean;
  exitCode: number | null;
  stdout: string;
  stderr: string;
  parsedResult: unknown | null;
}

/// ログストリーム
export type LogStream = 'stdout' | 'stderr';

/// リアルタイムログイベント
export interface LogEvent {
  toolName: string;
  line: string;
  stream: LogStream;
  timestamp: string;
}

/// ツール実行状態
export type ToolStatus = 'running' | 'completed' | 'failed' | 'cancelled';

/// ツール実行状態イベント
export interface ToolStatusEvent {
  toolName: string;
  status: ToolStatus;
  result: ToolResult | null;
}

/// ログエントリ
export interface LogEntry {
  line: string;
  stream: LogStream;
  timestamp: string;
}

